### PR TITLE
Renamed GRASS GIS locations to projects

### PR DESCRIPTION
**Pull request by anikaweinmann**
Renamed GRASS GIS locations to projects from GRASS GIS version 8.4:

- add decorator for deprecated locations
- add endpoints for all projects-endpoint-urls
- Add tests for G8.3 and G8.4

### CODE REVIEW

File: docker/actinia-core-alpine/Dockerfile [hunk 1]
```diff
@@ -16,12 +16,12 @@ ENV GISBASE ""
 COPY --from=grass /usr/local/bin/grass /usr/local/bin/grass
 COPY --from=grass /usr/local/grass* /usr/local/grass/
 RUN ln -s /usr/local/grass "$(grass --config path)"
-RUN grass --tmp-location EPSG:4326 --exec g.version -rge && \
+RUN grass --tmp-project EPSG:4326 --exec g.version -rge && \
     pdal --version && \
```
Comment (mmacata) [inline docker/actinia-core-alpine/Dockerfile:20]: Also above the FROM image should be changed to g84

File: src/actinia_core/processing/actinia_processing/ephemeral/download_cache_management.py [hunk 2]
```diff
@@ -42,7 +42,10 @@
 __copyright__ = (
     "Copyright 2016-2022, Sören Gebbert and mundialis GmbH & Co. KG"
 )
-__maintainer__ = "mundialis"
+__maintainer__ = "mundialis GmbH & Co. KG"
+__email__ = "info@mundialis.de"
+__maintainer__ = "mundialis GmbH & Co. KG"
+__email__ = "info@mundialis.de"
```
Comment (mmacata) [inline src/actinia_core/processing/actinia_processing/ephemeral/download_cache_management.py:49]:
```suggestion
__maintainer__ = "mundialis GmbH & Co. KG"
__email__ = "info@mundialis.de"

```

File: tests/test_noauth.py [hunk 3]
```diff
@@ -22,7 +22,7 @@
 #######
 
 """
-Tests: Location test case
+Tests: Projct test case
```
Comment (mmacata) [inline tests/test_noauth.py:25]:
```suggestion
Tests: Project test case
```

File: tests/test_webhook.py [hunk 4]
```diff
@@ -144,7 +145,7 @@ def poll_job(self, resp_data):
     #         time.sleep(3)
     #     tm = Template(json_dumps(pc))
     #     rv = self.server.post(
-    #         URL_PREFIX + "/locations/nc_spm_08/processing_async_export",
+    #         URL_PREFIX + "/projects/nc_spm_08/processing_async_export",
```
Comment (mmacata) [inline tests/test_webhook.py:148]: Maybe add test for both until GRASS GIS 9 ?

File: tests/test_webhook.py [hunk 5]
```diff
@@ -157,7 +158,7 @@ def test_finished_webhook_retries(self):
         """
         tm = Template(json_dumps(pc))
         rv = self.server.post(
-            URL_PREFIX + "/locations/nc_spm_08/processing_async_export",
+            URL_PREFIX + "/projects/nc_spm_08/processing_async_export",
```
Comment (mmacata) [inline tests/test_webhook.py:161]: Same here - Maybe add test for both until GRASS GIS 9 ?

File: tests/test_strds_raster_renderer.py [hunk 6]
```diff
@@ -34,31 +34,27 @@
...
-location = "nc_spm_08"
+project = "nc_spm_08"
 strds_mapset = "modis_lst"
-strds_url = URL_PREFIX + "/locations/%(location)s/mapsets/%(mapset)s/strds" % {
-    "location": location,
-    "mapset": strds_mapset,
-}
+# strds_url = (
+#     f"{URL_PREFIX}/PROJECT_URL_PART/{project}/mapsets/{strds_mapset}/strds"
+# )
```
Comment (mmacata) [inline tests/test_strds_raster_renderer.py:48]: Why is this outcommented? Is this not needed?

File: src/actinia_core/rest/resource_management.py [hunk 7]
```diff
@@ -352,21 +353,22 @@ def _create_ResourceDataContainer_for_resumption(
             is None
         ):
             return None, None, None
-        location = re.findall(r"locations\/(.*?)\/", post_url)[0]
+        # TODO project project
```
Comment (mmacata) [inline src/actinia_core/rest/resource_management.py:356]: What does this TODO mean?

File: tests/test_strds_raster_renderer.py (hunk 6 above)
Comment (anikaweinmann) [reply tests/test_strds_raster_renderer.py:48]: Yes it is not used

File: src/actinia_core/testsuite.py [hunk 8]
```diff
@@ -451,7 +480,8 @@ def create_vector_layer(self, location, mapset, vector, region, parameter):
         )
         self.waitAsyncStatusAssertHTTP(
             rv,
-            headers=self.admin_auth_header,
+            headers=self.user_auth_header,
+            # headers=self.admin_auth_header,
```
Comment (mmacata) [inline src/actinia_core/testsuite.py:484]: Could this lead to errors?

Comment (anikaweinmann) [reply src/actinia_core/testsuite.py:484]: It has not lead to errors by other tests

File: src/actinia_core/rest/resource_management.py (hunk 7 above)
Comment (anikaweinmann) [reply src/actinia_core/rest/resource_management.py]: Das war nur die Erinnerung für mich. Habe gerade die aktuelle Version hochgeladen

File: tests/test_webhook.py (hunk 5 above)
Comment (anikaweinmann) [reply tests/test_webhook.py]: Ist schon geändert. War leider noch nicht commited, da ich die Tests verschoben hatte

File: tests/test_webhook.py (hunk 4 above)
Comment (anikaweinmann) [reply tests/test_webhook.py]: Ist schon geändert. War leider noch nicht commited, da ich die Tests verschoben hatte

File: tests/test_job_resumption.py [hunk 9]
```diff
@@ -4,7 +4,7 @@
 # performance processing of geographical data that uses GRASS GIS for
 # computational tasks. For details, see https://actinia.mundialis.de/
 #
-# Copyright (c) 2016-2018 Sören Gebbert and mundialis GmbH & Co. KG
+# Copyright (c) 2021-20224 Sören Gebbert and mundialis GmbH & Co. KG
```
Comment (mmacata) [inline tests/test_job_resumption.py:7]:
```suggestion
# Copyright (c) 2021-2024 Sören Gebbert and mundialis GmbH & Co. KG
```

File: tests/test_strds_raster_renderer.py [hunk 10]
```diff
@@ -34,31 +34,24 @@
...
 __email__ = "soerengebbert@googlemail.com"
 
-location = "nc_spm_08"
+project = "nc_spm_08"
 strds_mapset = "modis_lst"
-strds_url = URL_PREFIX + "/locations/%(location)s/mapsets/%(mapset)s/strds" % {
-    "location": location,
-    "mapset": strds_mapset,
-}
 strds_data = "LST_Day_monthly"
```
Comment (mmacata) [inline tests/test_strds_raster_renderer.py:46]: Is this used in this file?

File: src/actinia_core/rest/project_management.py [hunk 11]
```diff
@@ -120,14 +122,17 @@ def get(self):
...
                     ):
-                        locations.append(dir)
-        if locations:
+                        projects.append(dir)
+        if projects:
+            param = {"status": "success"}
+            grass_version_s = G_VERSION["version"]
+            grass_version = [int(item) for item in grass_version_s.split(".")[:2]]
+            if grass_version >= [8, 4]:
+                param["projects"] = projects
```
Comment (mmacata) [inline src/actinia_core/rest/project_management.py:131]: This might break with `LocationListResponseModel`

Comment (mmacata) [reply src/actinia_core/rest/project_management.py:131]: Fine if this PR stays open as long as it takes to also adjust actinia-api but I would suggest to do adjustments in a separate PR.

File: tests/test_job_resumption.py [hunk 12]
```diff
@@ -364,8 +364,9 @@ def test_saved_interim_results(self):
         """Test if the interim results are not saved correctly"""
         step = 4
         tpl = Template(json_dumps(process_chain_1))
+        # self.server.get(f"{URL_PREFIX}/{self.project_url_part}/nc_spm_08/mapsets",headers=self.admin_auth_header,content_type="application/json")
```
Comment (mmacata) [inline tests/test_job_resumption.py:367]: This is just a test for debugging, right?

File: src/actinia_core/core/common/api_logger.py [hunk 13]
```diff
@@ -110,7 +110,9 @@ def add_entry(self, user_id, http_request):
 
         """
         api_info = {
-            "endpoint": http_request.endpoint,
+            # For depricated location endpoints remove "_locations" from
```
Comment (mmacata) [inline src/actinia_core/core/common/api_logger.py:113]:
```suggestion
            # For deprecated location endpoints remove "_locations" from
```

File: src/actinia_core/core/common/api_logger.py [hunk 14]
```diff
@@ -110,7 +110,9 @@ def add_entry(self, user_id, http_request):
...
+            # endpoint class name
+            "endpoint": http_request.endpoint.replace("_locations", ""),
```
Comment (mmacata) [inline src/actinia_core/core/common/api_logger.py:115]: Why is this necessary?

File: src/actinia_core/rest/base/resource_base.py [hunk 15]
```diff
@@ -183,7 +183,9 @@ def __init__(self, resource_id=None, iteration=None, post_url=None):
 
         # Put API information in the response for later accounting
         kwargs = {
-            "endpoint": request.endpoint,
+            # For depricated location endpoints remove "_locations" from
```
Comment (mmacata) [inline src/actinia_core/rest/base/resource_base.py:186]:
```suggestion
            # For deprecated location endpoints remove "_locations" from
```

File: src/actinia_core/rest/base/resource_base.py [hunk 16]
```diff
@@ -183,7 +183,9 @@ def __init__(self, resource_id=None, iteration=None, post_url=None):
...
+            # endpoint class name
+            "endpoint": request.endpoint.replace("_locations", ""),
```
Comment (mmacata) [inline src/actinia_core/rest/base/resource_base.py:188]: And here? I guess for the same reason :upside_down_face:

File: src/actinia_core/core/common/api_logger.py (hunk 14 above)
Comment (anikaweinmann) [reply src/actinia_core/core/common/api_logger.py:115]: It is needed because I can't add two urls for one endpoint. The endpoint is inside the actinia response and is used for the job jesumption.

File: src/actinia_core/rest/base/resource_base.py (hunk 16 above)
Comment (anikaweinmann) [reply src/actinia_core/rest/base/resource_base.py:188]: yes

File: src/actinia_core/endpoints.py [hunk 17]
```diff
@@ -129,164 +127,302 @@
...
+__email__ = "info@mundialis.de"
 
 
-def create_core_endpoints():
-    # Locationmanagement
-    flask_api.add_resource(ListLocationsResource, "/locations")
+def get_endpoint_class_name(endpoint_class, projects_url_part="projects"):
+    endpoint_class_name = endpoint_class.__name__.lower()
+    if "project" in endpoint_class_name:
+        name = endpoint_class_name.replace("project", projects_url_part[:-1])
```
Comment (mmacata) [inline src/actinia_core/endpoints.py:139]: Doesn't this replace "project" with "project"? Why is this needed?

File: src/actinia_core/core/common/api_logger.py (hunk 14 above)
Comment (mmacata) [reply src/actinia_core/core/common/api_logger.py:115]: Ah I see. Would it be possible to use the variable `projects_url_part`? (Probably not because of circular reference)

File: src/actinia_core/rest/base/resource_base.py (hunk 16 above)
Comment (mmacata) [reply src/actinia_core/rest/base/resource_base.py:188]: Commented above

**Review by mmacata — APPROVED** [review]
:partying_face:

File: src/actinia_core/core/common/api_logger.py (hunk 14 above)
Comment (mmacata) [reply src/actinia_core/core/common/api_logger.py:115]: As discussed we leave it like this to be able to find and remove it with GRASS GIS 9

File: src/actinia_core/rest/base/resource_base.py (hunk 16 above)
Comment (mmacata) [reply src/actinia_core/rest/base/resource_base.py:188]: and above

File: src/actinia_core/endpoints.py [hunk 18]
```diff
@@ -129,164 +127,302 @@
...
+    elif projects_url_part != "projects":
+        name = f"{endpoint_class_name}_{projects_url_part}"
+    else:
+        name = endpoint_class_name
+    return name
```
Comment (mmacata) [inline src/actinia_core/endpoints.py:144]:
```suggestion
def get_endpoint_class_name(endpoint_class, projects_url_part="projects"):
    endpoint_class_name = endpoint_class.__name__.lower()
    if projects_url_part != "projects":
        name = f"{endpoint_class_name}_{projects_url_part}"
    else:
        name = endpoint_class_name
    return name
```
